python 3.8.0

bayandin opened the pull request:
I update PR description from time to time to track the progress of python 3.8 migration.

There is a branch for tracking progress: https://github.com/bayandin/homebrew-core/tree/test-python-3.8.0

---

- [x] **breezy** ([breezy.log](https://github.com/Homebrew/homebrew-core/files/3864844/breezy.log)) https://github.com/Homebrew/homebrew-core/pull/46941

```
    running build_ext
    skipping 'breezy/_simple_set_pyx.c' Cython extension (up-to-date)
    building 'breezy._simple_set_pyx' extension
    creating build/temp.macosx-10.15-x86_64-3.8
    creating build/temp.macosx-10.15-x86_64-3.8/breezy
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Versions/8.5/Headers -Ibreezy -I/usr/local/opt/python/Frameworks/Python.framework/Versions/3.8/include/python3.8 -c breezy/_simple_set_pyx.c -o build/temp.macosx-10.15-x86_64-3.8/breezy/_simple_set_pyx.o
    breezy/_simple_set_pyx.c:6808:270: error: too many arguments to function call, expected 15, have 16
      __pyx_codeobj__9 = (PyObject*)__Pyx_PyCode_New(3, 0, 5, 0, CO_OPTIMIZED|CO_NEWLOCALS, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__8, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_stringsource, __pyx_n_s_pyx_unpickle__SimpleSet_iterat, 1, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__9)) __PYX_ERR(1, 1, __pyx_L1_error)
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
    breezy/_simple_set_pyx.c:318:82: note: expanded from macro '__Pyx_PyCode_New'
              PyCode_New(a, 0, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
              ~~~~~~~~~~                                                             ^~~~
    /usr/local/opt/python/Frameworks/Python.framework/Versions/3.8/include/python3.8/code.h:122:12: note: 'PyCode_New' declared here
    PyAPI_FUNC(PyCodeObject *) PyCode_New(
               ^
    breezy/_simple_set_pyx.c:8619:9: error: too many arguments to function call, expected 15, have 16
            __pyx_empty_bytes  /*PyObject *lnotab*/
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    breezy/_simple_set_pyx.c:318:82: note: expanded from macro '__Pyx_PyCode_New'
              PyCode_New(a, 0, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
              ~~~~~~~~~~                                                             ^~~~
    /usr/local/opt/python/Frameworks/Python.framework/Versions/3.8/include/python3.8/code.h:122:12: note: 'PyCode_New' declared here
    PyAPI_FUNC(PyCodeObject *) PyCode_New(
               ^
    2 errors generated.

      Cannot build extension "breezy._simple_set_pyx".
      Use "build_ext --allow-python-fallback" to use slower python implementations instead.

    error: command 'clang' failed with exit status 1
    Running setup.py install for breezy: finished with status 'error'
``` 

- [x] **gpgme** ([gpgme.log](https://github.com/Homebrew/homebrew-core/files/3864848/gpgme.log)) https://github.com/Homebrew/homebrew-core/pull/45337#issuecomment-555810021
```
Testing gpgme
/usr/bin/sandbox-exec -f /private/tmp/homebrew20191119-88892-17ta5r4.sb ruby -W0 -I $LOAD_PATH -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gpgme.rb --verbose
==> /usr/local/Cellar/gpgme/1.13.1_1/bin/gpgme-tool --lib-version
==> python2.7 -c import gpg; print gpg.version.versionstr
1.13.1
==> python3 -c import gpg; print(gpg.version.versionstr)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'gpg'
Error: gpgme: failed
```

- [x] **graph-tool** ([graph-tool.log](https://github.com/Homebrew/homebrew-core/files/3870361/graph-tool.log)) https://github.com/Homebrew/homebrew-core/pull/46987
```
In file included from /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/__string:57:
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:2494:5: error: static_assert failed due to requirement '__is_forward_iterator<boost::adj_list<unsigned long>::edge_iterator>::value' "std::max_element requires a ForwardIterator"
    static_assert(__is_forward_iterator<_ForwardIterator>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/graph/graph_adjacency.hh:452:26: note: in instantiation of function template specialization 'std::__1::max_element<boost::adj_list<unsigned long>::edge_iterator, (lambda at ../../../src/graph/graph_adjacency.hh:453:38)>' requested here
        auto iter = std::max_element(erange.first, erange.second,
                         ^
../../../src/graph/graph.hh:109:33: note: in instantiation of member function 'boost::adj_list<unsigned long>::shrink_to_fit' requested here
    void shrink_to_fit() { _mg->shrink_to_fit(); }
                                ^
1 error generated.
make[3]: *** [graph_betweenness.lo] Error 1
```

- [x] **openimageio** ([openimageio.log](https://github.com/Homebrew/homebrew-core/files/3865540/openimageio.log)) https://github.com/Homebrew/homebrew-core/pull/45337#issuecomment-555749877
```
In file included from /tmp/openimageio-20191119-78942-7ji6wt/oiio-Release-2.0.12/src/python/py_imageinput.cpp:31:
/tmp/openimageio-20191119-78942-7ji6wt/oiio-Release-2.0.12/src/python/py_oiio.h:41:10: fatal error: 'Python.h' file not found
#include <Python.h>
         ^~~~~~~~~~
1 error generated.
make[2]: *** [src/python/CMakeFiles/PyOpenImageIO.dir/py_imageinput.cpp.o] Error 1
make[1]: *** [src/python/CMakeFiles/PyOpenImageIO.dir/all] Error 2
```

- [x] **weechat** ([weechat.log](https://github.com/Homebrew/homebrew-core/files/3865551/weechat.log)) https://github.com/Homebrew/homebrew-core/pull/45337#issuecomment-555749877
```
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/plugins/python/python.so] Error 1
make[1]: *** [src/plugins/python/CMakeFiles/python.dir/all] Error 2
```

- [x] **wxpython** ([wxpython.log](https://github.com/Homebrew/homebrew-core/files/3865709/wxpython.log)) https://github.com/Homebrew/homebrew-core/pull/46952
```
   [853/875] Compiling sip/cpp/sip_webkitwxWebKitBeforeLoadEvent.cpp
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:48: error: unknown class name 'wxWebKitStateChangedEvent'; did you mean 'sipwxWebKitStateChangedEvent'?
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~
                                                   sipwxWebKitStateChangedEvent
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:7: note: 'sipwxWebKitStateChangedEvent' declared here
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
          ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:46: error: base class has incomplete type
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
                                         ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:7: note: definition of 'sipwxWebKitStateChangedEvent' is not complete until the closing '}'
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
          ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:25:43: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
        sipwxWebKitStateChangedEvent(const  ::wxWebKitStateChangedEvent&);
                                            ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                                              wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:46:81: error: member initializer 'wxWebKitStateChangedEvent' does not name a non-static data member or base class
    sipwxWebKitStateChangedEvent::sipwxWebKitStateChangedEvent( ::wxWindow*win):  ::wxWebKitStateChangedEvent(win), sipPySelf(SIP_NULLPTR)
                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:51:69: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
    sipwxWebKitStateChangedEvent::sipwxWebKitStateChangedEvent(const  ::wxWebKitStateChangedEvent& a0):  ::wxWebKitStateChangedEvent(a0), sipPySelf(SIP_NULLPTR)
                                                                      ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                                                                        wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:51:104: error: member initializer 'wxWebKitStateChangedEvent' does not name a non-static data member or base class
    sipwxWebKitStateChangedEvent::sipwxWebKitStateChangedEvent(const  ::wxWebKitStateChangedEvent& a0):  ::wxWebKitStateChangedEvent(a0), sipPySelf(SIP_NULLPTR)
                                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:69:19: error: no member named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'sipwxWebKitStateChangedEvent'?
            return  ::wxWebKitStateChangedEvent::GetEventCategory();
                    ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                      sipwxWebKitStateChangedEvent
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:7: note: 'sipwxWebKitStateChangedEvent' declared here
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
          ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:84:19: error: no member named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'sipwxWebKitStateChangedEvent'?
            return  ::wxWebKitStateChangedEvent::Clone();
                    ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                      sipwxWebKitStateChangedEvent
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:7: note: 'sipwxWebKitStateChangedEvent' declared here
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
          ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:100:12: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
             ::wxWebKitStateChangedEvent *sipCpp;
             ~~^~~~~~~~~~~~~~~~~~~~~~~~~
               wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:109:30: error: no member named 'GetState' in 'wxPaletteChangedEvent'
                sipRes = sipCpp->GetState();
                         ~~~~~~  ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:135:12: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
             ::wxWebKitStateChangedEvent *sipCpp;
             ~~^~~~~~~~~~~~~~~~~~~~~~~~~
               wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:146:21: error: no member named 'SetState' in 'wxPaletteChangedEvent'
                sipCpp->SetState(state);
                ~~~~~~  ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:172:12: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
             ::wxWebKitStateChangedEvent *sipCpp;
             ~~^~~~~~~~~~~~~~~~~~~~~~~~~
               wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:181:46: error: no member named 'GetURL' in 'wxPaletteChangedEvent'
                sipRes = new  ::wxString(sipCpp->GetURL());
                                         ~~~~~~  ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:208:12: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
             ::wxWebKitStateChangedEvent *sipCpp;
             ~~^~~~~~~~~~~~~~~~~~~~~~~~~
               wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:219:21: error: no member named 'SetURL' in 'wxPaletteChangedEvent'
                sipCpp->SetURL(*url);
                ~~~~~~  ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:247:18: error: no type named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'wxPaletteChangedEvent'?
            const  ::wxWebKitStateChangedEvent *sipCpp;
                   ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                     wxPaletteChangedEvent
    ../../../ext/wxWidgets/include/wx/event.h:2861:24: note: 'wxPaletteChangedEvent' declared here
    class WXDLLIMPEXP_CORE wxPaletteChangedEvent : public wxEvent
                           ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:256:50: error: no member named 'wxWebKitStateChangedEvent' in the global namespace; did you mean 'sipwxWebKitStateChangedEvent'?
                sipRes = (sipSelfWasArg ? sipCpp-> ::wxWebKitStateChangedEvent::Clone() : sipCpp->Clone());
                                                   ~~^~~~~~~~~~~~~~~~~~~~~~~~~
                                                     sipwxWebKitStateChangedEvent
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:21:7: note: 'sipwxWebKitStateChangedEvent' declared here
    class sipwxWebKitStateChangedEvent : public  ::wxWebKitStateChangedEvent
          ^
    ../../../sip/cpp/sip_webkitwxWebKitStateChangedEvent.cpp:256:77: error: 'sipwxWebKitStateChangedEvent::Clone' is not a member of class 'const ::wxPaletteChangedEvent'
                sipRes = (sipSelfWasArg ? sipCpp-> ::wxWebKitStateChangedEvent::Clone() : sipCpp->Clone());
                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    20 errors generated.
```

* [x] **ansible-lint**: https://github.com/Homebrew/homebrew-core/pull/47030 ( upstream issue https://github.com/ansible/ansible-lint/issues/590)
* [x] **salt**: https://github.com/Homebrew/homebrew-core/pull/45895 (upstream issue https://github.com/saltstack/salt/issues/55084)
* [x] **mrboom**: https://github.com/Homebrew/homebrew-core/pull/47004
* [x] Also, see https://github.com/Homebrew/homebrew-core/issues/46728

methane commented:
Would you add `--enable-optimization` configure option?

From Python 3.8, the option doesn't take too long time to build.
See https://bugs.python.org/issue36044

fxcoudert commented:
> Would you add --enable-optimization configure option?

Given this is not a default option, probably not, unless it is done in upstream (“official”) binaries.

fxcoudert commented:
@bayandin Major version update, this is going to be a fun ride…

- `pip` update in `python` should not be a separate commit
- formulas that hardcode `3.7`: `ola xapian zim`

methane commented:
I think official binaries uses `--enable-optimizations` and `--with-lto`.  But I'll confirm it.

bayandin commented:
@fxcoudert 

> @bayandin Major version update, this is going to be a fun ride…
> 
> * `pip` update in `python` should not be a separate commit
> * formulas that hardcode `3.7`: `ola xapian zim`

No problem! I'm on it

fxcoudert commented:
Looking at first test results, all `python`-dependent formulas will need a revision bump 😭

bayandin commented:
Pushed a bunch of revision bumps + fixes for `ola`, `xapian` and `zim`.

Jenkins output from the previous build (just in case):
[python3.8-highsierra.txt](https://github.com/Homebrew/homebrew-core/files/3728177/python3.8-highsierra.txt)
[python3.8-mojave.txt](https://github.com/Homebrew/homebrew-core/files/3728178/python3.8-mojave.txt)

bayandin commented:
Have pushed remaining revision bumps for formulas which depend on python and failed and weren't fixed yet:

```
alot
arcade-learning-environment
aws-google-auth
awslogs
b2-tools
biogeme
boost-python3
buku
bumpversion
carla
certbot
codespell
credstash
cython
duplicity
euler-py
fb-client
fbi-servefiles
gdal
gdcm
genometools
git-cola
git-revise
gitfs
gobject-introspection
goolabs
graph-tool
gst-python
itstool
jinja2-cli
jrnl
keepkey-agent
lensfun
libxml2
livestreamer
mitmproxy
mps-youtube
nanopb-generator
numpy
opencv
opencv@3
openimageio
pastebinit
py3cairo
pybind11
pygobject3
pyqt
pyside
pyvim
qscintilla2
scipy
thefuck
theharvester
trezor-agent
twine-pypi
urh
vapoursynth
wxpython
yapf
```

bayandin commented:
@BrewTestBot test this please

fxcoudert commented:
@bayandin it will run when the CI has a free slot, we have to wait…

Noctem commented:
`--without-gcc` is not a recognized option. Only causes a warning, but might as well remove it.

> Would you add `--enable-optimizations` and `--with-lto` configure option?

`--with-lto` apparently requires llvm-ar which as far as I'm aware is not available in macOS' toolchain, so brewed llvm would presumably be required to build. `--enable-optimizations` seems like a good idea to me though.

Noctem commented:
Here are a couple changes I needed to make to dependencies to get them to compile against 3.8:
023c4582e4a133283da36d1cd24d19b2d1f552ba
b240b11be24827f226c025bd5bcb03768cbc834e 

VapourSynth was also failing to link so I added `-L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.8 -lpython3.8` to `LDFLAGS`, but I don't think that should be applied to every VapourSynth component (nor should it need to be supplied manually), so there should be a better fix.

methane commented:
It seems offcial macOS binary doesn't use both of `--enable-optimizations` and `--with-lto`.

But these options have significant performance gains. See https://bugs.python.org/issue38477 for a recent example.

fxcoudert commented:
> It seems offcial macOS binary doesn't use both of --enable-optimizations and --with-lto

So we'll stick to not using them.

> But these options have significant performance gains.

If it is safe, upstream should enable it by default. It does lead to significant build-time increase (30 minutes instead of 2 minutes, from what I see), and we do have users who build from source.

methane commented:
> If it is safe, upstream should enable it by default

It is not safe because we don't know much about the environment.  PGO and LTO were unstable in the old GCC and Python may be built on such platforms.  That's why we don't enable it by default in `Makefile`.

On the other hand, we can assume compiler versions on macOS.  I don't know why official Python binary doesn't use them.  I agree that the official binary should use it too.

lithammer commented:
Related discussion on the python-dev mailing list: https://mail.python.org/archives/list/python-dev@python.org/thread/EILILECNTLTW4VCBCPW37R4QRU7ZBDEU/#VUA2AG6J2QG2GK52DZ4VKZMRUTKJ4DWO

methane commented:
> Related discussion about the flags on the python-dev mailing list: https://mail.python.org/archives/list/python-dev@python.org/thread/EILILECNTLTW4VCBCPW37R4QRU7ZBDEU/#VUA2AG6J2QG2GK52DZ4VKZMRUTKJ4DWO

Official binary doesn't use PGO + LTO because it is built on macOS 10.9 and these option is not robust on such old build tools.
I think there is no reason to not using it in case of Homebrew.  Macports, Ubuntu, RedHat, etc... uses these options.

fxcoudert commented:
@bayandin please be careful not to force-push too often, otherwise we are exhausting the CI nodes by running this job over and over again (right now, I just killed one CI run because our whole CI was running this twice)

bayandin commented:
@fxcoudert ah, thanks for letting me know. CI queuing looks a bit intransparent for me now.

bertonha commented:
Python 3.8 still needs the "openssl@1.1" as dependency ?
The standard "openssl" is not enough ?

I think the formula `pre-commit` will also need a revision bump

bayandin commented:
@bertonha 
>  Python 3.8 still needs the "openssl@1.1" as dependency ?
The standard "openssl" is not enough ?

According to https://github.com/Homebrew/homebrew-core/pull/971 `openssl@1.1` — is the "standard" openssl, and `openssl` is the previous one version (1.0.2)

> I think the formula pre-commit will also need a revision bump

Thanks! I'll check it and will update if necessary.

fxcoudert commented:
Hi @bayandin we're going to try and clear the CI backlog for a few hours… meanwhile, please take a look at the failure from the first CI job, see what you can fix. And we can rerun another batch tomorrow, when we have merged a few other PRs.

I'm sorry testing on this is so difficult, right now these large-scale PRs (major updates for python, boost, icu4c, openssl, gettext, readline) really reach the limits of our CI strategy.

bayandin commented:
It seems CI has stuck, I suspect the problem in building `pytouhou`. I've pushed a fix for `pytouhou` (and for several other formulas).
@fxcoudert could you please restart the job

bayandin commented:
@fxcoudert I've updated PR description and have mentioned all known problems, also extracted update of resources to a separate PR (https://github.com/Homebrew/homebrew-core/pull/45629). 
Could you please stop all running CI builds for this PR and sort it outs mentioned issues.
Thanks!

utkonos commented:
`--enable-optimizations` and `--with-lto` are for production environments. I would hope that someone isn't just installing something from homebrew in a production environment. I'd leave this up to the user if they know what it does and then want to compile with options like this. If `--with-lto` has an additional package dependency, perhaps include some checks in the formula that installs that dep if the user compiles and installs with that option?

Not sure about the best way to handle it.

aaronliu0130 commented:
Just checked this out. Keep up the good work!

JonZeolla commented:
Looks like there's a merge conflict to address now

bayandin commented:
@fxcoudert could you please kill the Catalina build: https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/50498/

It seems there is a problem with `brew audit --online ...` on Catalina and it hangs forever (I've reported it here: https://github.com/Homebrew/brew/issues/6648).

SMillerDev commented:
@bayandin could you rebase again?

bayandin commented:
@SMillerDev sure gonna do it later after checking/fixing (some of) failed tests. Don't want to push to frequently because it takes a day to rebuild everything on CI. And merge conflicts here is the least of the problems. Is that ok?

SMillerDev commented:
Sure thing

fxcoudert commented:
- Need revision bump: `doitlive nbdime vim@7.4`
- `pyqt` is currently failing to build from source 😢  (related to `sip` trouble)
- `gpgme` fails test with:  `ModuleNotFoundError: No module named 'gpg'`
- `dvc` fails to build with: `ModuleNotFoundError: No module named 'Cython'`
- `ansible-lint` fails to build (preexisting failure)
- `breezy` fails to build

CI is taking very long because it's trying over and over again to build `pyqt` from source (due to failure above). I'm killing the build because it could last for many days, at this speed. Once the items above are fixed (especially `pyqt`), we can give it another shot.

cclauss commented:
Given all the conflicts above, can this get re-based please?

fxcoudert commented:
@cclauss as explained in my comment above (https://github.com/Homebrew/homebrew-core/pull/45337#issuecomment-549020356), rebasing is the least of the issues here. We need to fix `sip` and `pyqt` before we can do anything else. Help is welcome!

cclauss commented:
@fxcoudert I am really struggling with how to help here.  I have done a [Python formula upgrade in the past](https://github.com/Homebrew/homebrew-core/pull/2415) but perhaps _micro_ releases are trivial compared to _minor_ releases.  Based on your advise above, I went to #45822 but that also had conflicts.  Now that PR has been closed in favor of #44964 which covers the same items also has conflicts.   Where exactly is assistance required?

bayandin commented:
Hey @cclauss, I'll resolve the conflicts later today and will rerun CI job.
Also, there is a list of failed formulas (see PR description), currently the following formulas are requiring assistance (for fixing build or test process): 
```
- breezy
- dvc
- gdal
- gpgme
- graph-tool
- openimageio
- pdal
- weechat
- wxpython
```

Thanks!

cclauss commented:
Thanks, I will dig in there...

A hint for future task lists in GitHub so we can tick things off as we go...
https://help.github.com/en/github/managing-your-work-on-github/about-task-lists

bayandin commented:
Just for the note.
When I initially was working on this PR, there was a problem with building `PyNaCl` python package and I have changed it to use `libsodium` from homebrew.
It seems now it is fixed in [`Xcode 11.2`](https://github.com/jedisct1/libsodium/issues/881#issuecomment-549049136), but Xcode wasn't updated on homebrews CI yet (see https://github.com/Homebrew/brew/pull/6702).
After Xcode update on CI, it will be needed to revert these changes

fxcoudert commented:
@cclauss yes, Python 3.x updates really are major releases, and pretty much all our Python formulas need to be rebuilt. This is always a long process, as incompatibilities arise. I've spent some hours fixing pyqt (at least to some level… I didn't manage to get the latest sip working). Help could be useful in getting sip working (https://github.com/Homebrew/homebrew-core/pull/44964) or understanding how to fix `vtk` (see https://gitlab.kitware.com/vtk/vtk/issues/17724)

@bayandin thanks for rebasing, now that `pyqt` is merged. We still have `vtk` not building with Xcode 11 (Mojave and Catalina)… not sure how much it will impact the CI.

Bo98 commented:
> Help could be useful in understanding how to fix `vtk` (see https://gitlab.kitware.com/vtk/vtk/issues/17724)

It's due to particular part of their code technically not following the C++ standard, but libc++ has only added assertions recently to detect such cases. libstdc++ hasn't yet afaik.

I have it fixed:

```
$ brew install --build-from-source vtk
==> Downloading https://www.vtk.org/files/release/8.2/VTK-8.2.0.tar.gz
Already downloaded: /Users/bo/Library/Caches/Homebrew/downloads/bfbea40437b17f2734e32acfebf51e9d00c1dd8a34d63ac7410c77fdb46883e9--VTK-8.2.0.tar.gz
==> Downloading https://gitlab.kitware.com/Bo98/vtk/commit/464152efaf9cb4bd0f8d5
######################################################################## 100.0%
Warning: Cannot verify integrity of 9d0c3bbde64f2c3ad23badf1c42bb63f4d353a804ce5e75cc6aaf284f364271c--464152efaf9cb4bd0f8d527e344b6c10246a2b25.diff
A checksum was not provided for this resource
For your reference the SHA-256 is: 8165fa4ed10654603d019cf1f4f7a613a894a8f105ad5d27b82e4a7f8d3f748c
==> Patching
==> Applying 464152efaf9cb4bd0f8d527e344b6c10246a2b25.diff
patching file Infovis/BoostGraphAlgorithms/vtkBoostGraphAdapter.h
Hunk #3 succeeded at 244 with fuzz 1 (offset 1 line).
Hunk #4 succeeded at 280 (offset 1 line).
Hunk #5 succeeded at 308 (offset 1 line).
Hunk #6 succeeded at 328 (offset 1 line).
Hunk #7 succeeded at 344 (offset 1 line).
Hunk #8 succeeded at 381 (offset 1 line).
Hunk #9 succeeded at 397 (offset 1 line).
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG
==> make
==> make install
Warning: vtk dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
🍺  /usr/local/Cellar/vtk/8.2.0_4: 4,260 files, 131.2MB, built in 24 minutes 41 seconds
```

and will submit the patch upstream shortly, with details on what was the issue if anyone is interested. Not sure if I followed their preferred code style or not but I'll see what the response is.

Bo98 commented:
https://gitlab.kitware.com/vtk/vtk/merge_requests/6178

bayandin commented:
@fxcoudert could you please kill [Catalina build](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/51635/version=catalina/): `No space left on device`. And it seems [High Sierra build](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/51635/version=highsierra/console) has hung on `brew test --verbose mrboom` (the last message was 6.5 hours ago).

Bo98 commented:
`adios2` needs a rev bump.

SMillerDev commented:
Killed all builds

fxcoudert commented:
VTK issue with Xcode 11 will be fixed by https://github.com/Homebrew/homebrew-core/pull/46706

According to @Bo98 it has Python 3.8 compat patches that will be needed here, as well.

Bo98 commented:
VTK patch (to be applied after #46706 is merged):

```
patch do
  url "https://gitlab.kitware.com/vtk/vtk/commit/257b9d7b18d5f3db3fe099dc18f230e23f7dfbab.diff"
  sha256 "572c06a4ba279a133bfdcf0190fec2eff5f330fa85ad6a2a0b0f6dfdea01ca69"
end
```

Disclaimer that I have _not_ tested this patch, but it is included upstream. If there's any difficulties building even with this patch, ping me and I will investigate!

Bo98 commented:
I'm also looking into a couple of the other build issues. The bullet issue has been fixed (#46652). The MySQL and Protobuf conflict has been fixed (#46682). The fix for `adios2` is mentioned above.

bayandin commented:
@Bo98 amazing work!

I've bumped `adios2`, `mpi4py` and `jupyterlab` locally, will push it along with a patch for VTK just after https://github.com/Homebrew/homebrew-core/pull/46706

Bo98 commented:
Awesome. I'll ping you when #46706 is merged.

Bo98 commented:
Just noticed the VTK error in the last build is not what I expected:

```
FILE STRINGS file
  "/usr/local/opt/python/Frameworks/Python.framework/Versions/3.8/include/python3.8m/patchlevel.h"
  cannot be read.
```

@bayandin Can you confirm whether this file exists?

Bo98 commented:
@bayandin #46706 is now merged. Though I'd still check up on the above error before starting a full build.

bayandin commented:
> @bayandin #46706 is now merged. Though I'd still check up on the above error before starting a full build.

Thanks! I've added the patch and fixed that problem ([`python3.8m` doesn't exist for Python 3.8](https://docs.python.org/3/whatsnew/3.8.html#build-and-c-api-changes)). 
Rebased the branch and have pushed it🤞

Bo98 commented:
Thanks, let’s see how it goes.

fxcoudert commented:
Only one emoji can adequately describe this PR: 🎢

bayandin commented:
I have Deja vu (the same problems as [3 days ago](https://github.com/Homebrew/homebrew-core/pull/45337#issuecomment-552837918) with High Sierra and Mojave builds ):
- High Sierra build has hung on [`brew test --verbose mrboom`](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/51811/version=highsierra/console)
- Mojave build [ran out of disc space](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/51811/version=mojave/console)

Catalina build surprisingly going great so far (PLEASE DONT KILL IT)

Bo98 commented:
mrboom seems to be a recurring problem on High Sierra. Not sure if anyone is able to reproduce the hang locally.

ModdedGamers commented:
I'd be happy to help with this. Do you have a giant list, or is it the ones at the top?

bayandin commented:
Hey @ModdedGamers, there is a list of problems mentioned in https://github.com/Homebrew/homebrew-core/issues/46728, also from the previous build there were problems  with the following formulas (either with building or with testing):

```
- breezy
- dvc
- gdal
- gpgme
- graph-tool
- openimageio
- pdal
- weechat
- wxpython
```

bayandin commented:
It seems Catalina build has stuck on [`brew install --verbose --build-bottle binwalk`](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/51811/version=catalina/console)
@fxcoudert could you please kill it.

Bo98 commented:
Have we covered `sip` in the rev bumps? I think I saw a pyqt sip issue earlier.

Bo98 commented:
In regards to #46728, its worth noting that they are non-blockers for this pull request. That said, we're close to resolving all of them. I'm looking at `infer` right now and `cucumber-cpp` may be blocked due to a CLT SDK issue (or would could switch to use brew ruby rather than system ruby - not sure what the current direction forward is for that).

bayandin commented:
> Have we covered sip in the rev bumps? I think I saw a pyqt sip issue earlier.

`sip` itself doesn't require a revision bump (I've double-checked it locally), but there is a problem with `pyqt`: it gets built, but the test fails (as far as I remember it was like this since `pyqt` 5.13.2 (https://github.com/Homebrew/homebrew-core/pull/46598):

```
13:37:53 ==> brew test pyqt --verbose
13:37:58 ==> FAILED
13:37:58 Testing pyqt
13:37:58 /usr/bin/sandbox-exec -f /private/tmp/homebrew20191115-72494-rku9p.sb ruby -W0 -I $LOAD_PATH -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pyqt.rb --verbose
13:37:58 ==> /usr/local/Cellar/pyqt/5.13.2_1/bin/pyuic5 --version
13:37:58 Traceback (most recent call last):
13:37:58   File "/usr/local/Cellar/python/3.8.0/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 192, in _run_module_as_main
13:37:58     return _run_code(code, main_globals, None,
13:37:58   File "/usr/local/Cellar/python/3.8.0/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 85, in _run_code
13:37:58     exec(code, run_globals)
13:37:58   File "/usr/local/lib/python3.8/site-packages/PyQt5/uic/pyuic.py", line 26, in <module>
13:37:58     from PyQt5 import QtCore
13:37:58 ModuleNotFoundError: No module named 'PyQt5.sip'
13:37:58 Error: pyqt: failed
```

There is a bit more explanation in https://github.com/Homebrew/homebrew-core/pull/44964#issuecomment-548010500

bayandin commented:
> In regards to #46728, its worth noting that they are non-blockers for this pull request. That said, we're close to resolving all of them. I'm looking at `infer` right now and `cucumber-cpp` may be blocked due to a CLT SDK issue (or would could switch to use brew ruby rather than system ruby - not sure what the current direction forward is for that).

@Bo98 great work!

Bo98 commented:
Doesn't sip install to `lib/python3.x/site-packages`?

In regards to `--sip-module`, that's already done: https://github.com/Homebrew/homebrew-core/commit/110b0f77bab3cf3fbe1eb72edf8deadf387d9a4f

bayandin commented:
> Doesn't sip install to `lib/python3.x/site-packages`?
> 
> In regards to `--sip-module`, that's already done: [110b0f7](https://github.com/Homebrew/homebrew-core/commit/110b0f77bab3cf3fbe1eb72edf8deadf387d9a4f)

Actually yes, you're absolutely right. I've bumped the revision and will push it later

Bo98 commented:
Yeah, there's no point pushing at the moment. There's a recurring mrboom issue on High Sierra and a recurring disk space issue on Mojave anyway.

PetrPy commented:
I am sorry to disturb, I know almost nothing about brew inner workings, but I have a question: Does anybody know what went wrong, given that Python 3.8 brings almost solely backwards-compatible non-breaking changes? Or is it an issue with new versions of some other dependencies? Thanks in advance for a patient answer :)

The-Compiler commented:
> given that Python 3.8 brings almost solely backwards-compatible non-breaking changes?

I'm "only" a Homebrew users as well, but the maintainer of a [moderately complex/popular](https://qutebrowser.org/) Python project - this is pretty much ideal thinking, and not true in either theory or practice :wink:

In theory: Previously deprecated features and APIs [are removed](https://docs.python.org/3/whatsnew/3.8.html#api-and-feature-removals) with every Python upgrade (and new [deprecation warnings introduced](https://docs.python.org/3/whatsnew/3.8.html#deprecated))

In practice: A Python release brings around 1.5 years of changes (between Python 3.7 and 3.8: 4335 commits,  2213 files changed, 266603 insertions, 111696 deletions). Even in one project, you're likely going to adapt in some way. Even if the changes are subtle, there's quite a big chance they have *some* kind of effect you need to adjust to. On a scale of hundreds of projects (like with Homebrews), there are going to be a lot of problems and backwards-incompatible changes.

bayandin commented:
Hey @PetrPy, thanks for the question

I think we have a set of coincidence of changes in different places which led us to this delay in Python update, here is some of them from my point of view:
- Python 3.8 ABI changes like [this](https://docs.python.org/3/whatsnew/3.8.html#debug-build-uses-the-same-abi-as-release-build), and C API changes like [this](https://docs.python.org/3/whatsnew/3.8.html#build-and-c-api-changes) both requires reviewing formulas with C-extensions
- [Removing deprecated `linux_distribution`](https://bugs.python.org/issue28167) method has affected some formulas
- A lot of formulas depends on python and have to be rebuilt/retested, more formulas depend on depended on python formulas and so on. The whole build process takes more that one day and has uncovered non-related to python build problems (you can check them https://github.com/Homebrew/homebrew-core/issues/46728)
- Also, macOS Catalina and Xcode 11 releases weren't smooth I'd like to see them which hardened testing a lot

PetrPy commented:
Thanks for your explanation, guys. I have clearly underestimated the scaling of relatively small changes in a big project like this. I keep my fingers crossed!

Bo98 commented:
@bayandin Do you have a summary of the precise issues with the broken formulae?

I'm aware of the upstream issues with ansible-lint (fixed in 4.1.1 currently in pre-release) and salt (does not work with Homebrew OpenSSL as it is keg only - it only tries to find OpenSSL at runtime too). And I know about gpgme not working for some reason (would be good to see what the Cellar or bottle looks like for a broken build).

gdal seemed fine in the last Catalina build and we didn't quite get to see the others which is what I'm asking about.

bayandin commented:
> @bayandin Do you have a summary of the precise issues with the broken formulae?

I'm collecting it. Will update you when results are ready

bayandin commented:
@Bo98 I have updated the PR description with failed formulas from older builds (I reproduced them locally on Catalina).

Bo98 commented:
Awesome. Many thanks for doing that! I'll have a look over.

Bo98 commented:
Breezy 3.0.2 has been released today which should fix that build.

henryiii commented:
Several of the affected formulas are probably due to a change in CPython 3.8: Python no longer makes it too easy to link to libpython [See notes here](https://docs.python.org/3/whatsnew/3.8.html#debug-build-uses-the-same-abi-as-release-build). You were never supposed to do it for normal modules, but `python3-config --libs` made it a bit to easy. `--embed` will add it if it is really needed.

Bo98 commented:
* **breezy** - update as above. I'll open a PR for that.
* **gpgme** and **graph-tool** - run the following and note its difference between 3.7 and 3.8:
```py
import distutils.sysconfig
distutils.sysconfig.get_config_var('LINKFORSHARED')
```
* **openimageio** - remove the m suffix: https://github.com/Homebrew/homebrew-core/blob/11732b1101405d586f8db4b455a24176aeb983e0/Formula/openimageio.rb#L54
* **weechat** - apply patch: https://patch-diff.githubusercontent.com/raw/weechat/weechat/pull/1421.patch?full_index=1
* **wxPython** - bit of a mess this one, but I'm going to try update it to see if that helps.

Bo98 commented:
Ok, I tried updating wxPython and it failed with the same error locally so nothing to do with Python 3.8 but I'll need to figure out why it's now broken.

bayandin commented:
> * **gpgme** and **graph-tool** - run the following and note its difference between 3.7 and 3.8:
> 
> ```python
> import distutils.sysconfig
> distutils.sysconfig.get_config_var('LINKFORSHARED')
> ```

Here they are:
```
-Wl,-stack_size,1000000  -framework CoreFoundation /usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/Python
```
and 
```
-Wl,-stack_size,1000000  -framework CoreFoundation Python.framework/Versions/3.8/Python
```

Bo98 commented:
Yeah, that's a problem. Need to try and get that path absolute again.

bayandin commented:
> Yeah, that's a problem. Need to try and get that path absolute again.

It looks like we need to reimplement https://github.com/Homebrew/homebrew-core/pull/22103

Bo98 commented:
Good find! Looks like that patch is still there but seemingly nonfunctional now.

Bo98 commented:
Could it possibly be `_sysconfigdata__darwin_darwin.py` now? Given:

```py
def _get_sysconfigdata_name():
    return os.environ.get('_PYTHON_SYSCONFIGDATA_NAME',
        '_sysconfigdata_{abi}_{platform}_{multiarch}'.format(
        abi=sys.abiflags,
        platform=sys.platform,
        multiarch=getattr(sys.implementation, '_multiarch', ''),
    ))
```
(https://github.com/python/cpython/blob/3.8/Lib/sysconfig.py#L346-L352)

Bo98 commented:
wxPython was an absolute pain in the arse but I've (hopefully) dealt with it now: #46952.

bayandin commented:
@Bo98 I've applied all the fixes you suggested. All is good only `graph-tool` started to fail with another error:
```
In file included from /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/__string:57:
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:2494:5: error: static_assert failed due to requirement '__is_forward_iterator<boost::adj_list<unsigned long>::edge_iterator>::value' "std::max_element requires a ForwardIterator"
    static_assert(__is_forward_iterator<_ForwardIterator>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/graph/graph_adjacency.hh:452:26: note: in instantiation of function template specialization 'std::__1::max_element<boost::adj_list<unsigned long>::edge_iterator, (lambda at ../../../src/graph/graph_adjacency.hh:453:38)>' requested here
        auto iter = std::max_element(erange.first, erange.second,
                         ^
../../../src/graph/graph.hh:109:33: note: in instantiation of member function 'boost::adj_list<unsigned long>::shrink_to_fit' requested here
    void shrink_to_fit() { _mg->shrink_to_fit(); }
                                ^
1 error generated.
make[3]: *** [graph_betweenness.lo] Error 1
```

Thanks for all the work you're doing!

Bo98 commented:
Ha - I am getting déjà vu after dealing with VTK! Exact same error as there.

I will have a look. Thank you.

Bo98 commented:
#46987

Bo98 commented:
Small update here: my pull requests for breezy, wxPython and graph-tool are merged.

I currently have #47004 open to help with mrboom getting stuck on High Sierra. It passed the tests the first time it went through, and hopefully will the second time - so that issue should be resolved soon.

As for the last two:
* ansible-lint is a blocker. I'm not sure when 4.1.1 is released, but https://github.com/ansible/ansible-lint/commit/c91d23b2f82b4efb540168132842b243eb2d8b0d.patch?full_index=1 is a possible (untested by myself) patch to apply in the meantime.
* salt as far as I know is a test failure and is not caused by Python 3.8. The test currently fails on 3.7. The problem is that upstream, with the way it searches at runtime, would effectively need to have a list of known paths - including `/usr/local/opt/openssl@1.1` - to fix it. So basically add special cases for Homebrew (and other package managers) - which may start an argument of who's responsible for that.
  - An inreplace of `find_library("crypto")` with the result of `Formula["openssl@1.1"].opt_lib/"libcrypto.dylib"` is a possible approach. We'll need to discuss a way forward and such direction can be taken in a separate pull request.

bayandin commented:
Thanks @bo98 for keeping us updated!

> my pull requests for breezy, wxPython and graph-tool are merged.

I'm testing these changes locally against Python 3.8. `breezy` and `graph-tool` are ok. `wxPython` is in progress.

> * ansible-lint is a blocker. I'm not sure when 4.1.1 is released, but https://github.com/ansible/ansible-lint/commit/c91d23b2f82b4efb540168132842b243eb2d8b0d.patch?full_index=1 is a possible (untested by myself) patch to apply in the meantime.

Not caused by Python 3.8 as well, the problem is compatibility with the latest `setuptools` version, and I believe will fail with Python 3.7 as well. Will try to apply the patch in a separate PR.
Also, have asked about 4.1.1 release https://github.com/ansible/ansible-lint/issues/590#issuecomment-557113246, 

> salt as far as I know is a test failure and is not caused by Python 3.8. The test currently fails on 3.7. The problem is that upstream, with the way it searches at runtime, would effectively need to have a list of known paths - including `/usr/local/opt/openssl@1.1` - to fix it. So basically add special cases for Homebrew (and other package managers) - which may start an argument of who's responsible for that.
> * An inreplace of `find_library("crypto")` with the result of `Formula["openssl@1.1"].opt_lib/"libcrypto.dylib"` is a possible approach. We'll need to discuss a way forward and such direction can be taken in a separate pull request.

The idea with using `Formula["openssl@1.1"].opt_lib/"libcrypto.dylib"` looks ok to me and this kind of solution already suggested in the original issue (https://github.com/saltstack/salt/issues/55084#issuecomment-552510505), will test it in https://github.com/Homebrew/homebrew-core/pull/45895 probably and could be discussed there as well.

Bo98 commented:
> Not caused by Python 3.8 as well, the problem is compatibility with the latest `setuptools` version, and I believe will fail with Python 3.7 as well.

Yeah, good point - I see Python 3.7 already is using 41.6.0 which I think is a problematic version for ansible-lint.

> will test it in #45895 probably and could be discussed there as well.

I agree. Let's try it out there.

Bo98 commented:
As for the non-blockers, #46728, I hope to finish almost all of that list off over the next couple days, depending on how feedback on the Infer pull request goes. One exception is `cucumber-cpp` which cannot be fixed at the moment due to an issue with the SDK setup. That's being partially tracked under #46973 and I'll need to reraise that the SDKs haven't been fixed yet - the focus just now is the CI backlog.

bayandin commented:
I believe a bunch of new Python 3 formulas are coming in https://github.com/Homebrew/homebrew-core/issues/47050

Bo98 commented:
Oh that's a fair point. Getting this merged first might be a good idea. Salt is the last blocker for now I think but that'll be merged soon.

Unless other pull requests are merged in the meantime, only the ones you changed, dnsviz and osc, are new for now. The ansible ones are removals and ledger had the Python part stripped out.

fxcoudert commented:
> Once the build starts, we should ideally tag any pull request on formula depending on Python 3 as "do not merge".

We can try that. If that is not sufficient, what we do for some very big pull requests like this one, is that we can lock `homebrew-core` completely (preventing any merge) for one or two days.

bayandin commented:
> dnsviz and osc, are new for today

Yep, have bumped revision for these two and for `uwsgi`.

bayandin commented:
Fasten your seatbelt, we're doing another try!

Bo98 commented:
Few questions:
* apache-arrow
* cassandra
* z3

bayandin commented:
> Few questions:
> 
> * apache-arrow
> * cassandra
> * z3
> 
> (this is just me looking through open PRs which depend on Python 3)

You're right! Bumped revisions and added python tests for these formulas.

The whole CI stopped because of `No available formula with the name "grib-api"` in `eccodes`. `grib-api` was deleted in https://github.com/Homebrew/homebrew-core/pull/46924.
Have fixed this, and pushed the changes.

From the short run (https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/52326/):
- [ ] **abyss** https://github.com/bcgsc/abyss/issues/305
- [ ] **alure** and **dromeaudio** https://github.com/Homebrew/homebrew-core/issues/46728
- [ ] **aom**
```
04:20:46 ==> FAILED
04:20:47 Testing aom
04:20:47 /usr/bin/sandbox-exec -f /private/tmp/homebrew20191123-96092-1lzb2nj.sb ruby -W0 -I $LOAD_PATH -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/aom.rb --verbose
04:20:47 ==> Downloading https://media.xiph.org/video/derf/y4m/bus_qcif_15fps.y4m
04:20:47 /usr/bin/curl -q --globoff --show-error --user-agent Homebrew/2.1.16-127-ga1939cf\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.1\)\ curl/7.64.1 --fail --silent --location --remote-time --continue-at 0 --output /Users/brew/Library/Caches/Homebrew/downloads/ced756f9d1f53c598f4f2b5c3676f9a313d4102ac938a50a36d66ea7343e22bd--bus_qcif_15fps.y4m.incomplete https://media.xiph.org/video/derf/y4m/bus_qcif_15fps.y4m
04:20:47 curl: (60) SSL certificate problem: certificate has expired
```
- [x] **mercurial** — missed revision bump during the rebase. Fixed
- [ ] **bazelisk** on High Sierra: the same problem as in https://github.com/Homebrew/homebrew-core/pull/46983

bayandin commented:
Bad news, I just find out that haven't bump revisions for the following formulae which requires a `python` as a runtime dependency (53):
```
afflib
anjuta
aurora-cli
bear
bento4
bind
choose
cryptominisat
ddgr
distcc
dnstwist
emscripten
evince
fpp
geos
gimme-aws-creds
glib
global
googler
gr-osmosdr
gradio
grc
gtk-doc
hashpump
ispc
lcm
liblouis
libproxy
mftrace
micropython
mitie
mm-common
nicovideo-dl
open-babel
opencolorio
otf2
percol
ponysay
protobuf@3.6
redex
shocco
simgrid
ssh-audit
sshconfigfs
sslmate
suricata
tbb
uhd
unoconv
vit
watchman
whatmp3
yosys
```

Not sure how I missed all of these, probably their tests never failed for new Python 3.8.

fxcoudert commented:
I'd say let's keep CI running for now, see what it uncovers already.

Bo98 commented:
alure and dromeaudio are doomed to fail on the CI because they deal with audio. It's difficult writing meaningful tests for those without using audio.

I've having similar issues with fluid-synth. There's a critical bug fix that I tried opening a pull request for (the previous version is simply broken) but even the dumb `--version` check that was there before broke as they forgot to update the version. My attempts to write another test that also worked on the CI ultimately failed and the pull request is kinda just stuck in limbo at the moment.

So yeah, just have to ignore that for now.

Not really sure what's happening with `aom`. Might have been a temporary thing.

Bo98 commented:
I've updated #46728 with what can be seen up to now.

In addition (excluding those your list of those to revision bump):

* dafny - known to fail if z3 receives any type of revision or version bump - see #46991. The fix is known but it's failing tests on High Sierra on rebuild.
* uhd - you haven't revision bumped this yet but I have since been informed that this won't build on Catalina. I'm experimenting in #47101 but I don't really know what the problem is yet.

I also have major concerns about disk space. The Mojave node started this build with 75GB of disk space left and now has just 3GB remaining...

basictheprogram commented:
Is there any grunt work I can help on? Not a lot of experience with contributing to home brew but I'm a python developer by profession and would like to help if I can.

aaronliu0130 commented:
> Is there any grunt work I can help on? Not a lot of experience with contributing to home brew but I'm a python developer by profession and would like to help if I can.

Me too, though I am a superuser rather than a python developer.

Bo98 commented:
I think the main issue right now is disk space to build all these formulae on the CI. Will have to cancel the current build because of that.

Apart from that we're getting close and a lot of the hard work has already been done.

@bayandin apache-arrow: `ModuleNotFoundError: No module named 'pyarrow'`

bayandin commented:
> @bayandin apache-arrow: `ModuleNotFoundError: No module named 'pyarrow'`

My bad attempt to improve the test. Fixed now.

> I think the main issue right now is disk space to build all these formulae on the CI. Will have to cancel the current build because of that.

Is there a plan what to do with it?

Bo98 commented:
> Is there a plan what to do with it?

I've not heard, but I haven't heard much of anything today. I'll find out because we certainly need to do something.

In terms of formulae, I suppose there's one blocker still to fix and that's #47101. Gave up on it yesterday - I'll need to get a Catalina setup to look into it.

cclauss commented:
[50 days since Python 3.8 shipped.](https://www.python.org/downloads/release/python-380)  How are things progressing?

bayandin commented:
> 50 days since Python 3.8 shipped.

Unfortunately, we can't use this information to help us with migration ¯\\\_(ツ)\_/¯

However, as we find out in this PR It's impossible to migrate everything (all `python` dependant formulae) at once (due to CI limitation and amount of work that should be done), so we're working on adding `python@3.8` and gradual migration. There is a bunch of tight-connected to each other formulae which have to be migrated together, and this is the first and the most challenging issue which we're trying to solve in https://github.com/Homebrew/homebrew-core/pull/47326 (I would say it is in good shape so far). 
When it's done, we're ready to migrate other independent formulae one-by-one.

michaelblyons commented:
Can an owner add the [python] label to this PR?

bertonha commented:
> 50 days since Python 3.8 shipped.

@bayandin I agree that kind of message does not help the migration, but I think we got on the wrong way to update to Python3.8.
My point is Python3.8 is way more important at least for me than most of the others packages, especially the ones failing and delaying this. e.g. VTK and a bunch of others.

Maybe if we had create a formula to `python@3.7` and linked all the "broken" formulas to it. And Update the formula `python` to Python3.8. We could already had python 3.8 and incrementally fixing formulas to run on python3.8.

This will also facilitate to parallel work on this issue.

SMillerDev commented:
> My point is Python3.8 is way more important at least for me than most of the others packages, especially the ones failing and delaying this. e.g. VTK and a bunch of others.

Unfortunately for you homebrew doesn't run for you, it runs to be useful to as many people as possible. Updating python forcefully because Christofer wants us to isn't really in line with that goal. If you want a python 3.8 version faster you're welcome to make a tap though.

> Maybe if we had create a formula to `python@3.7` and linked all the "broken" formulas to it. And Update the formula `python` to Python3.8. We could already had python 3.8 and incrementally fixing formulas to run on python3.8.

That is pretty much what is happening, except there is a pull request for a `python@3.8` formula. But ~5000 formula aren't checked in a day, and certainly don't have their conflicts resolved in a day.

bertonha commented:
> Unfortunately for you homebrew doesn't run for you, it runs to be useful to as many people as possible. Updating python forcefully because Christofer wants us to isn't really in line with that goal. If you want a python 3.8 version faster you're welcome to make a tap though.

@SMillerDev don't get me wrong. I love homebrew and it always worked really well, I am here just proposing a difference approach for big changes like this.

Usually big bang approses does not go to smoothly and create gigantic PR, like update python and all the dependencies in one-shot.

Again I am just proposing a incremental way to achieve those complexes upgrades.

utkonos commented:
> @SMillerDev don't get me wrong. I love homebrew and it always worked really well, I am here just proposing a difference approach for big changes like this.
> 
> Usually big bang approses does not go to smoothly and create gigantic PR, like update python and all the dependencies in one-shot.
> 
> Again I am just proposing a incremental way to achieve those complexes upgrades.

I just created a thread on the community forum about how to do python updates when there is a large cluster of dependencies that also need updating. I don't think the discussion should happen here on the PR. Traffic here should just be from from/for folks actually working on solving the PR. Extra traffic here probably just irritates the volunteers who are trying to solve the problem.

https://discourse.brew.sh/t/why-python-3-8-has-not-been-updated-yet/6057

iMichka commented:
`sslyze` probably needs to be added to this list (#47993).

bayandin commented:
Closed in favour of separate migration in #47274
